### ISSUE TITLE
refactor file/qml loading code

### ISSUE THREAD

**Issue by henrikrudstrom** [qmlweb/qmlweb#126]
Think we could clean up the structure a bit and make it easier to optimize and cache restructure the code relating to
* Resolving imports to files
* registering import paths
* retrieving files
* parsing files or loading parse trees / compiled qrc files
* etc

Advantages:
* Easier to implement efficient caching and other improvements (async file loading? etc)
* Unload file handling responsibility from QMLEngine.js

A start could be to move the import related code to the `qtcore/qrc.js` file and expose a simple api that can be used both by the qml engine and gulp-task.

**Comment by igosoft** [comment]
Indeed, it's one of most important subjects for qmlweb and it has influence on the qmlweb performance (response time).

**Comment by aspidites** [comment]
Would something like virtual-dom be relevant here?

**Comment by henrikrudstrom** [comment]
@aspidites sounds cool, ReaQt. i have no idea about the work involved to implement that, its a bit beyond what I had in mind with this issue.  
you could open an issue on it so we can disuss it and put it in the roadmap if its relevant.

**Comment by ChALkeR** [comment]
We need to finish the merges first, I think =).

**Comment by pavelvasev** [comment]
@henrikrudstrom @ChALkeR seems it already needs at least a small refactor. Today I discovered that current import state seems to be incomplete, because Component constructor doesn't call engine::loadImports. It is seen on the picture below -
![current import schema](http://win.lineact.com/sync_data/projects/clarity/pic1/out.svg)

**Comment by pavelvasev** [comment]
Update: the picture shown is incomplete without `createComponent` method. In turn that method calls `engine::loadImports`, so things get imported. But we need to coordinate 3 methods of importing currently present:

1. Import of JS files [performed in Component::finalizeImports, called by Component constructor]
2. Import of built-in components added with registerQmlType [in global::loadImports, called by Component constructor]
3. Import of qmldir files and import directories [engine::loadImports and createComponent /loadQMLTree]

**Comment by henrikrudstrom** [comment]
@pavelvasev Nice diagram. Hope we can reduce the amount of red lines :)

One simplification i can think of is if to store the registeredQmlTypes and the Parsed Qml types in the same list/structure. now the registred types are stored in `modules` in qml.js:
``` 
 modules[options.module] = {
      name:        options.name,
      versions:    options.versions,
      constructor: options.constructor
    };
```
From what i see the parsed qml types are stored as a map of `filename => tree` in the `qrc` object.
Qt.createComponent caches the components in `engine.components`

is there any reason `modules` and `engine.components` cannot be merged?

Also noticed that `engine.loadQMLTree()` and `Qt.createComponent()` seem to be pretty much the same...

**Comment by aspidites** [comment]
Sorry, I was more tossing out what might reduce unnecessary renders. I definitely think tackling open branches and solving import complexity is more important. virtual-dom and similar would make for nice greenfield side projects, but I don't think they are important enough to track as concrete tickets, yet. 

TL;DR - carry on, folks, nothing to see here from me!

**Comment by henrikrudstrom** [comment]
@aspidites but its still a good idea. Not sure if issues (with a label or milestone) is a good place to store good ideas... but would be good to have some way of catching them.
just noticed our newest startgazer @Swatinem is working on a virtual dom ;)

**Comment by Swatinem** [comment]
hah, that escalated quickly :-D My vdom experiments are ~2 years old. I’m now on the React bandwagon. But still skeptical especially concerning performance, etc.

But I do search for a truly *declarative* ui framework that tries to move things to compile time instead of runtime like react.

**Comment by henrikrudstrom** [comment]
you cant watch without being watched ;)

qml *declarative* alright, no idea where we stand performance wise, but i suspect not so good... yet.

**Comment by stephenmdangelo** [comment]
I have noticed that the global `constructors` Object is overwritten each time `global.loadImports` is called.  This results in only the imports from the last-loaded Component being available to `construct`. Is this related to this issue, or something different?
